src/simbricks/pci.cc

```diff
@@ -0,0 +1,735 @@
+        volatile union SimbricksProtoPcieH2D *msg = adapter.outAlloc();
+        volatile struct SimbricksProtoPcieH2DDevctrl *devctrl = &msg->devctrl;
+
+        devctrl->flags = 0;
+        if (intx_after)
+            devctrl->flags |= SIMBRICKS_PROTO_PCIE_CTRL_INTX_EN;
+        if (msi_after)
+            devctrl->flags |= SIMBRICKS_PROTO_PCIE_CTRL_MSI_EN;
+        if (msix_after)
+            devctrl->flags |= SIMBRICKS_PROTO_PCIE_CTRL_MSIX_EN;
+
+        adapter.outSend(msg, SIMBRICKS_PROTO_PCIE_H2D_MSG_DEVCTRL);
+    }
+
+    return t;
+}
+
+Device::DMACompl::DMACompl(Device *dev_, uint64_t id_, size_t bufsiz_,
+        enum ctype ty_, const std::string &name_)
+    : EventFunctionWrapper([this]{ done(); }, name_, true), dev(dev_), id(id_),
+    ty(ty_), buf(new uint8_t[bufsiz_]), bufsiz(bufsiz_)
+{
+}
+
+Device::DMACompl::~DMACompl()
+{
+    delete[] buf;
+}
+
+void
+Device::DMACompl::done()
+{
+    dev->dmaDone(*this);
+}
+
+void
+Device::dmaDone(DMACompl &comp)
+{
+    DPRINTF(SimBricksPci, "simbricks-pci: completed DMA id %u\n", comp.id);
+
+    if (comp.ty == DMACompl::READ) {
+        volatile union SimbricksProtoPcieH2D *msg = adapter.outAlloc();
+        volatile struct SimbricksProtoPcieH2DReadcomp *rc;
+        /* read completion */
+        rc = &msg->readcomp;
+        rc->req_id = comp.id;
+        memcpy((void *) rc->data, comp.buf, comp.bufsiz);
+        adapter.outSend(msg, SIMBRICKS_PROTO_PCIE_H2D_MSG_READCOMP);
+    } else if (comp.ty == DMACompl::WRITE) {
+        volatile union SimbricksProtoPcieH2D *msg = adapter.outAlloc();
+        volatile struct SimbricksProtoPcieH2DWritecomp *wc;
+        /* write completion */
+        wc = &msg->writecomp;
+        wc->req_id = comp.id;
+        adapter.outSend(msg, SIMBRICKS_PROTO_PCIE_H2D_MSG_WRITECOMP);
+    } else if (comp.ty == DMACompl::MSI) {
+        /* MSI interrupt */
+    } else {
+        panic("simbricks-pci: invalid completion");
+    }
+}
+
+void
+Device::handleInMsg(volatile union SimbricksProtoPcieD2H *msg)
+{
+    volatile struct SimbricksProtoPcieD2HRead *read;
+    volatile struct SimbricksProtoPcieD2HWrite *write;
+    volatile struct SimbricksProtoPcieD2HReadcomp *rc;
+    volatile struct SimbricksProtoPcieD2HWritecomp *wc;
+    volatile struct SimbricksProtoPcieD2HInterrupt *intr;
+    DMACompl *dc;
+    PciPioCompl *pc;
+    uint64_t rid, addr, len;
+    uint8_t ty;
+
+    ty = adapter.inType(msg);
+    switch (ty) {
+        case SIMBRICKS_PROTO_PCIE_D2H_MSG_READ:
+            /* Read */
+            read = &msg->read;
+
+            rid = read->req_id;
+            addr = read->offset;
+            len = read->len;
+            DPRINTF(SimBricksPci, "simbricks-pci: received DMA read id %u "
+                    "addr %x size %x\n", rid, addr, len);
+
+            dc = new DMACompl(this, rid, len, DMACompl::READ, name());
+            dmaRead(pciToDma(addr), len, dc, dc->buf, 0);
+            break;
+
+        case SIMBRICKS_PROTO_PCIE_D2H_MSG_WRITE:
```

Comment:
We are not handling incoming posted writes. I think we are not using them anywhere though. I used them in the JPEG decoder and VTA before but with the updated AXI library I actually need feedback on the write completion to not accept more than a limited number of pending writes at a time, similar to what a PCIe block on an FPGA does in practice.